File: cslavb/Csla/Core/PositionMap.vb
Imports System.Linq

Namespace Core

  Friend Class PositionMap(Of T As Core.IEditableBusinessObject)

    Private _map As Dictionary(Of T, Integer)

    'on rare occasions, we get duplicates.  Store those here
    Private _duplicatePositions As Dictionary(Of T, Integer) = New Dictionary(Of T, Integer) 'T is the item, the int is the position count (if there is a dupe)

    Private _list As IList(Of T)

    Public Sub New(ByVal list As IList(Of T))
      _list = list
      RebuildMap()
    End Sub

    Public Sub ClearMap()
      _map = New Dictionary(Of T, Integer)(_list.Count)
    End Sub

    Public Sub AddToDuplicates(ByVal item As T)
      If Not _duplicatePositions.ContainsKey(item) Then 'if there is not a duplicate entry already
        _duplicatePositions.Add(item, 1) 'number refers to # of dups
      Else
        _duplicatePositions(item) += 1
      End If
    End Sub

    Public Sub AddToMap(ByVal item As T)
      If Not _map.ContainsKey(item) Then
        _map.Add(item, _list.Count - 1)
      Else
        'its a duplicate - handle here
        Dim duplicateCount = _list.LongCount(Function(checkItem As T) ReferenceEquals(item, checkItem))

        If duplicateCount > 1 Then
          AddToDuplicates(item)
        End If

      End If
    End Sub

    Public Sub InsertIntoMap(ByVal item As T, ByVal position As Integer)
      If position = _list.Count - 1 Then
        AddToMap(item)
      Else
        If Not _map.ContainsKey(item) Then
          Dim i As Integer = _list.Count - 1
          Do While i > position
            _map(_list(i)) += 1
            i -= 1
          Loop
          _map.Add(item, position)
        End If
      End If
    End Sub

    Public Sub RemoveFromMap(ByVal item As T)
     
      If _duplicatePositions.ContainsKey(item) Then

        If _duplicatePositions(item) = 1 Then
          _duplicatePositions.Remove(item)
        Else
          _duplicatePositions(item) -= 1
        End If

        Return
      End If

      Dim oldPosition As Integer = PositionOf(item)
      If oldPosition = -1 Then
        Return
      End If

      _map.Remove(item)
      For i As Integer = oldPosition + 1 To _list.Count - 1
        _map(_list(i)) -= 1
      Next i
    End Sub

    Public Function PositionOf(ByVal item As T) As Integer
      If _map.ContainsKey(item) Then
        Return _map(item)
      Else
        Return -1
      End If
    End Function

    Public Sub RebuildMap()
      ClearMap()
      Dim i As Integer = 0
      For Each item As T In _list
        _map.Add(item, i)
        i += 1
      Next item

    End Sub

  End Class

End Namespace

File: cslacs/Csla/Core/PositionMap.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Csla.Core
{
  internal class PositionMap<T>
    where T : Core.IEditableBusinessObject
  {
    Dictionary<T, int> _map;
    //on rare occasions, we get duplicates.  Store those here
    Dictionary<T, int> _duplicatePositions = new Dictionary<T,int>(); //T is the item, the int is the position count (if there is a dupe)

    IList<T> _list;

    public PositionMap(IList<T> list)
    {
      _list = list;
      RebuildMap();
    }

    public void ClearMap()
    {
      _map = new Dictionary<T, int>(_list.Count);

    }

    public void AddToDuplicates(T item)
    {
      if (!_duplicatePositions.ContainsKey(item)) //if there is not a duplicate entry already
        _duplicatePositions.Add(item, 1); //number refers to # of dups
      else
        _duplicatePositions[item]++;
    }
    
    public void AddToMap(T item)
    {
      if (!_map.ContainsKey(item))
        _map.Add(item, _list.Count - 1);
      else //its a duplicate - handle here
      {
        var duplicateCount = _list.Count(checkItem => ReferenceEquals(item, checkItem));
        if (duplicateCount > 1) AddToDuplicates(item);
      }

    }

    public void InsertIntoMap(T item, int position)
    {
      if (position == _list.Count - 1)
      {
        AddToMap(item);
      }
      else
      {
        if (!_map.ContainsKey(item))
        {
          for (int i = _list.Count - 1; i > position; i--)
            _map[_list[i]]++;
          _map.Add(item, position);
        }
        else AddToDuplicates(item);
      }
    }

    public void RemoveFromMap(T item)
    {
      if (_duplicatePositions.ContainsKey(item))
      {
        if (_duplicatePositions[item] == 1)
          _duplicatePositions.Remove(item);
        else
          _duplicatePositions[item]--;
        return;
      }
      int oldPosition = PositionOf(item);
      if (oldPosition == -1) return;
      _map.Remove(item);
      for (int i = oldPosition + 1; i < _list.Count; i++)
        _map[_list[i]]--;
    }

    public int PositionOf(T item)
    {
      int result;
      if (_map.TryGetValue(item, out result))
        return result;
      else
        return -1;
    }

    public void RebuildMap()
    {
      ClearMap();
      int i = 0;
      foreach (T item in _list)
      {
        _map.Add(item, i);
        i++;
      }

    }
  }
}
